### ISSUE TITLE
Incorrect gradient for categorical distribution entropy

### ISSUE THREAD

**Issue by lwohlhart** [tensorflow/tensorflow#6717]
The **Categorical** distribution class provides an awesome **entropy** operator but apparently the **gradient** calculation w.r.t. the input operators **doesn't work**.

```python
logits = tf.Variable(initial_value=[[1., 2., 3.], [2., 5., 1.]])

probabilities = tf.nn.softmax(logits)
log_probabilities = tf.nn.log_softmax(logits)
entropy = - tf.reduce_sum(probabilities * log_probabilities, axis=-1)

# using the actual distribution would be nicer but gradients seem buggy
categorical_distribution = tf.contrib.distributions.Categorical(p=probabilities)
categorical_distribution_entropy = categorical_distribution.entropy()

# initialize
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)

# works
print(sess.run(entropy))
print(sess.run(tf.gradients(entropy, [logits])))

# apparently loses gradient information
print(sess.run(categorical_distribution_entropy))
print(sess.run(tf.gradients(categorical_distribution_entropy, [logits])))
```

In the outputs we see that the entropy calculation works but the gradients are somehow lost. This obviously also doesn't work when I try to maximize this entropy using any optimizer.
```
[ 0.83239555  0.27431309]
[array([[ 0.14181709,  0.14077036, -0.28258738],
       [ 0.13012242, -0.19513965,  0.06501719]], dtype=float32)]
[ 0.83239555  0.27431309]
[array([[ 0.,  0.,  0.],
       [ 0.,  0.,  0.]], dtype=float32)]
```

**Comment by yaroslavvb** [comment]
I suspect this a numerical stability issue. [Looking](https://gist.github.com/yaroslavvb/97504b8221a8529e7a51a50915206d68) at your "working" graph, it uses LogSoftmax and LogSoftmaxGrad nodes. Whereas in the categorical distribution, it computes SoftmaxGrad and then applies LogGrad on top of it. @ebrevdo

**Comment by ebrevdo** [comment]
Taking a look.

**Comment by ebrevdo** [comment]
For numerical stability, we use tf.nn.softmax_cross_entropy_with_logits to calculate the entropy.  I just found out that this operation does not perform backprop with respect to the "labels" argument (the exponentiated probabilities term of the entropy).  Thus the zeros.

Probably the best way to fix this is for us to implement the numerically stable equivalent of your reduce_sum.

**Comment by lwohlhart** [comment]
thank you for the fix !
i guess we can close this then
